### PR TITLE
WIP: Instance creation FloatingIPs from novaclient to neutronclient

### PR DESCRIPTION
**Pull request by junousi**
Novaclient floatingIP functionality is gone in queens. Replacing neutronclient functionality seems to be already in place in Newton. Proposing to merge these to master instead of queens/master.

### CODE REVIEW

File: files/nrpe/check_openstack.py [hunk 1]
```diff
@@ -286,23 +287,42 @@ def instance_destroy(self):
...
+            'floatingip': {
+                'port_id': target_id,
+            }
+        })
 
   def instance_detach_floating_ip(self):
     if hasattr(self, 'fip'):
       self.instance.remove_floating_ip(self.fip.ip)
-      self.nova.floating_ips.delete(self.fip)
+      self.neutron.delete_floatingip(self.fip['floatingip']['id'])
```
Comment (junousi) [inline files/nrpe/check_openstack.py:315]: From video review with team: This should be checked that it doesn't do anything dangerous if run as admin.

File: files/nrpe/check_openstack.py [hunk 2]
```diff
@@ -286,23 +287,42 @@ def instance_destroy(self):
...
 
   def floating_ip_delete(self):
     if hasattr(self, 'fip'):
-      self.nova.floating_ips.delete(self.fip)
+      self.neutron.delete_floatingip(self.fip['floatingip']['id'])
```
Comment (junousi) [inline files/nrpe/check_openstack.py:319]: From video review with team: This should be checked that it doesn't do anything dangerous if run as admin.

File: files/nrpe/check_openstack.py [hunk 3]
```diff
@@ -322,9 +342,9 @@ def delete_orphaned_instances(self):
       instance.delete()
 
   def delete_orphaned_floating_ips(self):
-    for tenant_ip in self.nova.floating_ips.list():
-      self.nova.floating_ips.delete(tenant_ip)
-    if len(self.nova.floating_ips.list()) != 0:
+    for tenant_ip in self.neutron.list_floatingips()['floatingips']:
+      self.neutron.delete_floatingip(tenant_ip['id'])
```
Comment (junousi) [inline files/nrpe/check_openstack.py:346]: From video review with team: This should be checked that it doesn't do anything dangerous if run as admin.

File: files/nrpe/check_openstack.py (hunk 1 above)
Comment (junousi) [reply files/nrpe/check_openstack.py:315]: - Here the delete operation is only targeted to self.fip
- In turn, self.fip is only ever assigned a value by create_floatingip(), nothing else in the code has a possibility to assign other values to it.

Based on this, I would deem these first two functions safe.

File: files/nrpe/check_openstack.py (hunk 2 above)
Comment (junousi) [reply files/nrpe/check_openstack.py:319]: See above

File: files/nrpe/check_openstack.py (hunk 3 above)
Comment (junousi) [reply files/nrpe/check_openstack.py:346]: Confirmed, when creating a network for nagiostest and running the test as nagiostest, the check removed IP from an unrelated canary. This should get a safeguard, even if right now we don't run it as admin user.

Comment (junousi) [reply files/nrpe/check_openstack.py]: Added a minimum viable safeguard.

File: files/nrpe/check_openstack.py [hunk 4]
```diff
@@ -322,10 +343,16 @@ def delete_orphaned_instances(self):
...
-      logging.warn('All floating IPs of instance creation test tenant were not deleted.')
+    # This used to be done with novaclient where it was enought to just list().
+    # With neutronclient one needs to be much more specific, otherwise if ran
+    # with an admin user there is a risk of getting all the Floating IPs.
+    own_project_id = self.session.get_project_id()
+    fip_lookup_params = { 'project_id': own_project_id, }
+    own_floating_ips = self.neutron.list_floatingips(**fip_lookup_params)['floatingips']
+    for project_ip in own_floating_ips:
+      self.neutron.delete_floatingip(project_ip['id'])
+    if len(own_floating_ips) != 0:
```
Comment (junousi) [inline files/nrpe/check_openstack.py:354]: whoops this part has broken logic, creating a fix

Comment (junousi) [reply files/nrpe/check_openstack.py]: updated

File: files/nrpe/check_openstack.py [hunk 5]
```diff
@@ -322,10 +343,15 @@ def delete_orphaned_instances(self):
...
-      logging.warn('All floating IPs of instance creation test tenant were not deleted.')
+    # This used to be done with novaclient where it was enought to just list().
+    # With neutronclient one needs to be much more specific, otherwise if ran
+    # with an admin user there is a risk of getting all the Floating IPs.
+    own_project_id = self.session.get_project_id()
+    fip_lookup_params = { 'project_id': own_project_id, }
+    for project_ip in self.neutron.list_floatingips(**fip_lookup_params)['floatingips']:
+      self.neutron.delete_floatingip(project_ip['id'])
+    if len(self.neutron.list_floatingips(**fip_lookup_params)['floatingips']) != 0:
+      logging.warn('All floating IPs of instance creation test project were not deleted.')
```
Comment (Aegil) [inline files/nrpe/check_openstack.py:354]: If the value is invalid does it still return all ips and then delete them anyhow.

eg if project_id gets renamed to tenant_id or borris_ip in your search. From a, would a keystone change break this for us.

Comment (junousi) [reply files/nrpe/check_openstack.py:354]: You mean

input: `{ 'project_id': own_project_id, }`
API changed and now expects: `{ 'borris_id': own_project_id, }`

?

That could be a risk. I could see about capping the result set. For example

- Select whichever is smaller, own project Floating IP quota or some reasonably low limit like 5
- If the result set exceeds the selected limit, halt and log an error

Comment (junousi) [reply files/nrpe/check_openstack.py]: In the end added a check based on Floating IP ownership.

Comment (Aegil) [reply files/nrpe/check_openstack.py]: I'm still thinking maybe the admin priv check is safer.

**Comment by junousi** [comment]
closing in favor of https://github.com/CSCfi/puppet-opsviewagent/pull/71

**Pull request closed without merging**